crates/accelerate/src/target_transpiler/nullable_index_map.rs

```diff
@@ -0,0 +1,455 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use ahash::RandomState;
+use indexmap::{
+    map::{IntoIter as BaseIntoIter, Iter as BaseIter, Keys as BaseKeys, Values as BaseValues},
+    IndexMap,
+};
+use pyo3::prelude::*;
+use pyo3::types::PyDict;
+use pyo3::IntoPy;
+use rustworkx_core::dictmap::InitWithHasher;
+use std::ops::Index;
+use std::{hash::Hash, mem::swap};
+
+type BaseMap<K, V> = IndexMap<K, V, RandomState>;
+
+///
+/// An `IndexMap`-like structure thet can be used when one of the keys can have a `None` value.
+///
+/// This structure is essentially a wrapper around the `IndexMap<K, V>` struct that allows the
+/// storage of `Option<K>` key values as `K`` and keep an extra slot reserved only for the
+/// `None` instance. There are some upsides to this including:
+///
+/// The ability to index using Option<&K> to index a specific key.
+/// Store keys as non option wrapped to obtain references to K instead of reference to Option<K>.
+///
+/// **Warning:** This is an experimental feature and should be used with care as it does not
+/// fully implement all the methods present in `IndexMap<K, V>` due to API limitations.
+#[derive(Debug, Clone)]
+pub struct NullableIndexMap<K, V>
```

Comment:
Well, with this one it's a little complicated. Even though it is currently being used only in `Target`, it might be usable for other things in the future where we have to store a nullable value inside of a dict-like structure. And maybe someone could build on top of it.

However, the main reason why I made it public is that if we call `Target.values` the value that is returned is an instance of `NullableIndexMap` and, if we wanted to avoid that, we'd have to reconstruct the `IndexMap` entirely. Of course, there might be a better way to approach this so I'm curious as to which ideas you might have.